src/ipnet.rs

```diff
@@ -0,0 +1,763 @@
+        match self {
+            IpNet::V4(inner) => inner.prefix(),
+            IpNet::V6(inner) => inner.prefix(),
+        }
+    }
+
+    /// Return `true` iff the subnet contains only the base address i.e. the
+    /// size is exactly one address.
+    pub fn is_host_net(&self) -> bool {
+        match self {
+            IpNet::V4(inner) => inner.is_host_net(),
+            IpNet::V6(inner) => inner.is_host_net(),
+        }
+    }
+
+    /// Return `true`` if the provided address is contained in self.
+    ///
+    /// This returns `false` if the address and the network are of different IP
+    /// families.
+    pub fn contains(&self, addr: IpAddr) -> bool {
+        match (self, addr) {
+            (IpNet::V4(net), IpAddr::V4(ip)) => net.contains(ip),
+            (IpNet::V6(net), IpAddr::V6(ip)) => net.contains(ip),
+            (_, _) => false,
+        }
+    }
+}
+
+impl From<Ipv4Net> for IpNet {
+    fn from(n: Ipv4Net) -> IpNet {
+        IpNet::V4(n)
+    }
+}
+
+impl From<Ipv6Net> for IpNet {
+    fn from(n: Ipv6Net) -> IpNet {
+        IpNet::V6(n)
+    }
+}
+
+impl std::fmt::Display for IpNet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            IpNet::V4(inner) => write!(f, "{}", inner),
+            IpNet::V6(inner) => write!(f, "{}", inner),
+        }
+    }
+}
+
+impl std::str::FromStr for IpNet {
+    type Err = IpNetParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let Some(ii) = s.find('/') else {
+            return Err(IpNetParseError::NoPrefix);
+        };
+
+        let addr_str = &s[..ii];
+        let prefix_str = &s[ii + 1..];
+
+        let prefix = prefix_str.parse().map_err(IpNetParseError::InvalidPrefix)?;
+        let addr = addr_str.parse().map_err(IpNetParseError::InvalidAddr)?;
+        IpNet::new(addr, prefix).map_err(IpNetParseError::PrefixValue)
+    }
+}
+
+#[cfg(feature = "schemars")]
+impl schemars::JsonSchema for IpNet {
+    fn schema_name() -> String {
+        "IpNet".to_string()
+    }
+
+    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        use crate::schema_util::label_schema;
+        schemars::schema::SchemaObject {
+            subschemas: Some(Box::new(schemars::schema::SubschemaValidation {
+                one_of: Some(vec![
+                    label_schema("v4", gen.subschema_for::<Ipv4Net>()),
+                    label_schema("v6", gen.subschema_for::<Ipv6Net>()),
+                ]),
+                ..Default::default()
+            })),
+            extensions: crate::schema_util::extension("IpNet", "0.1.0"),
+            ..Default::default()
+        }
+        .into()
+    }
+}
+
+/// The highest value for an IPv4 subnet prefix
+pub const IPV4_NET_PREFIX_MAX: u8 = 32;
+
+/// An IPv4 subnet
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Ipv4Net {
+    addr: Ipv4Addr,
+    prefix: u8,
```

Comment:
Calling this part the prefix is a bit awkward. If we were calling the whole data structure `Ipv4Prefix` or `Ipv4Subnet` this would be the `length`. In a world where we are just considering prefixes and subnets, I could also see calling what is now called `addr` the `prefix`, and calling this `length`. But that becomes awkward in the case where this data structure is to be used as an address/subnet-mask combo rather than a subnet or routing prefix. Maybe `width` works for both cases here?

If we were to have two fit-for-purpose data structures I think things would look like this

```rust
pub struct Ipv4Subnet {
    prefix: u32,
    length: u8,
}
```

```rust
pub struct Ipv4SubnetAddr {
    addr: Ipv4Addr,
    mask_length: u8,
}
```

And you'd have semantic differences like allowing non-zero bits in positions beyond the prefix length. But I'm not really sure this is worth separate data structures, and is somewhat contrary to the data structure proliferation-based pain we are trying to mitigate with this crate.

We could also just not name any of this and go with

```rust
pub struct Ipv4Net([u8;5])
```

and have different accessor methods that provide different values. The `addr` method would return an `IPv4Addr` composed on the leading 4 bytes independent of the mask, the `prefix` method would return an `Ipv4Addr` (or maybe a `u32`) with the bits beyond the prefix length masked out (which is subtly different than `network` as `prefix` would always be defined including for /31 and /32), `width` would return the 5th byte, etc.

Reply:
So: recall that these are non-`pub` members. I'm all for cleaning up the nomenclature, but I think that the external interface is going to be more important to nail. I saw dissensus among the various implementations... and just picked one even though it seemed imprecise.

With regard to using `[u8;5]` as the internal representation, I'd rather use an `Ipv4Addr` and `u8` for clarity and type safety. What advantage do you see with the former?

Reply:
Yeah totally agreed on nailing the external interface being the important part. The advantage I saw in `[u8;5]` was not applying a name to something that gives the wrong impression for some use cases of the data structure. If I'm using this data structure to represent `1.2.3/24` no part of that is really an address, it's just a prefix or a subnet with a specified length. But if these values are never really seen in the client code as they are private - and there are accessor methods that make semantic sense in the client code in which they are used like `prefix()` for the `1.2.3` part and `length()` or `width()` for the `/24`, that's great.